src/net.cpp

```diff
@@ -718,6 +720,127 @@ CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
     return msg;
 }
 
+int V2TransportDeserializer::Read(const char* pch, unsigned int bytes)
+{
+    if (!m_in_data) {
+        // copy data to temporary parsing buffer
+        const unsigned int remaining = CHACHA20_POLY1305_AEAD_AAD_LEN - m_hdr_pos;
+        const unsigned int copy_bytes = std::min(remaining, bytes);
+
+        memcpy(&vRecv[m_hdr_pos], pch, copy_bytes);
+        m_hdr_pos += copy_bytes;
+
+        // if AAD incomplete, exit
+        if (m_hdr_pos < CHACHA20_POLY1305_AEAD_AAD_LEN) {
+            return copy_bytes;
+        }
+
+        // we got the AAD bytes at this point (3 bytes encrypted packet length)
+        // we keep the sequence numbers unchanged at this point. Once the message is authenticated and decrypted, we increase the sequence numbers (or the aad_pos)
```

Comment:
Hmmm isn't the spec making the assumption you rely on some ordered transport protocol (TCP) ? If yes I think that's okay because you should be guarantee against network failure and peers may out-of-sync. But BIP should precise message shouldn't be resend with same sequence number?

Note IIRC Fibre is using UDP.

Reply:
I think we can make the assumptions that messages are processing in order and that the underlaying transport layer takes care of transmission failures. I don't think we need to add error/out-of-sequence detection in out message transport protocol.

Reply:
I agree it's good for now as an assumption. I think in the future there is valid use-cases where you want to send bitcoin traffic over some non-ordered communication channels like headers-over-radio and still benefit of encryption.

Reply:
I can't find this requirement in the BIP, not in "Packet Handling". Maybe you've already the change just withhold ?